proposals/0540-vision-for-layout-conformance-parity.md

```diff
@@ -0,0 +1,238 @@
+
+We should instead lean towards a future where components are shareable with web by default, while quarantining components which asume previous behaviors. Apart from better code-sharing, moving the bulk of the ecosystem towards tolerating W3C behavior unlocks new opportunities for [layout engine usage](#betting-on-yoga-alternative-choices). We must take steps to create a critical mass of usage in currently active projects. We can do this by making it easy to fix code which depends on non-conforming layout, and to provide tangible benefits for doing so.
+
+### Classic layout
+
+An app otherwise using strict layout may want to opt specific component trees into classic layout behavior. This can be useful both as a tool for incremental migration (starting by disabling strict layout where it causes differences), cases where code may be unable to be changed (e.g. third-party components), or as a tool to unblock fixing [new algorithm issues](#staging). `<StrictLayout>` then needs a negatinig counterpart `<ClassicLayout>` to allows opting out of a configurable subset of changes (especially useful for large migrations). 
+
+```jsx
+// Force fully classic rendering and disable migration hints
+<ClassicLayout>
+  <CruftView />
+<ClassicLayout>
+```
+```jsx
+// Disable a single behavior change and its migration hints
+<ClassicLayout quirks=['stale-basis-measurements']>
+  <CruftView /> 
+</ClassicLayout>
+```
+
+### Traceable styles
+
+It is not possible to statically reason about whether a given style will render differently under strict layout. We can instead rely on runtime feedback from the underlying layout engine. We can add an API to Yoga like `YGNodeAffectedByQuirk()`  to query if the engine made choices based on legacy behavior. E.g. on branching to `UseLegacyStretchBehaviour` we can calculate whether the different behavior will lead to a visible difference, and cache that on the yoga node.
+
+The Yoga Node attached to a Shadow Node is far enough away from the JS source defining a style that we need a method to correlate the two. This can be solved by a Babel transform, which inserts a correlation ID into each style, forwarded to the underlying Yoga Node. This ID can be mapped in the bundle to source location. This has a negative effect on bundle size, but should be possible to add to development mode bundles, as a debugging feature.
+
+The below shows a structure the babel transform could replicate:
+
+```jsx
+// Original stylesheet
+const styles = StyleSheet.create({
+  widget: {
+    flex: 1,
+  },
+});
+
+// After babel transform
+const styles = StyleSheet.create({
+  widget: StyleSheet.compose(
+    {
+      flex: 1,
+    },
+    {_creationOrigin: 123},
+  ),
+});
+
+// Style calculated by function
+<View style={calculateStyle()} />;
+
+// After babel transform
+<View style={StyleSheet.compose(calculateStyle(), {_applyOrigin: 456})} />;
```

Comment:
This would let the runtime provide a trace of where the native <View> is that passed an invalid style, but it wouldn't provide the definition site, is that right?

For example, in the Facebook app we have a layer of indirection called like <FacebookButton> that internally renders a `<View style={StyleSheet.compose(styles.view, props.childStyles)} />`. 

And then all the products call `<FacebookButton childStyles={styles.invalidStyle} />`. Would those all blame to the lines in `<FacebookButton>` or the callsites in product code? For debuggability, is there a way to mark the product callsite?

Reply:
I think it depends on how we would decide to handle compound (multiple in array), or composed styles.  The style of a ShadowNode can come from multiple contributing sources.

For the simpler case where there is a single style, the scheme in the example is to embed a single style-application-site and a single style-creation-site. Though if we associate one shadownode to a single style creation site, there is lossiness as styles are composed (some style takes precedence). So the right model might be to instead allow associating `n` style creation sites per style application site.